zerocopy-derive/src/repr.rs

```diff
@@ -6,202 +6,276 @@
 // This file may not be copied, modified, or distributed except according to
 // those terms.
 
-use core::fmt::{self, Display, Formatter};
-
-use proc_macro2::Ident;
-use quote::{ToTokens, TokenStreamExt as _};
+use core::{
+    convert::{Infallible, TryFrom},
+    num::NonZeroU64,
+};
 
-use {
-    proc_macro2::Span,
-    syn::punctuated::Punctuated,
-    syn::spanned::Spanned,
-    syn::token::Comma,
-    syn::{Attribute, DeriveInput, Error, LitInt, Meta},
+use proc_macro2::{Span, TokenStream};
+use quote::{quote_spanned, ToTokens, TokenStreamExt as _};
+use syn::{
+    punctuated::Punctuated, spanned::Spanned as _, token::Comma, Attribute, Error, LitInt, Meta,
+    MetaList,
 };
 
-pub(crate) struct Config<Repr: KindRepr> {
-    // A human-readable message describing what combinations of representations
-    // are allowed. This will be printed to the user if they use an invalid
-    // combination.
-    pub(crate) allowed_combinations_message: &'static str,
-    // Whether we're checking as part of `derive(Unaligned)`. If not, we can
-    // ignore `repr(align)`, which makes the code (and the list of valid repr
-    // combinations we have to enumerate) somewhat simpler. If we're checking
-    // for `Unaligned`, then in addition to checking against illegal
-    // combinations, we also check to see if there exists a `repr(align(N > 1))`
-    // attribute.
-    pub(crate) derive_unaligned: bool,
-    // Combinations which are valid for the trait.
-    pub(crate) allowed_combinations: &'static [&'static [Repr]],
-    // Combinations which are not valid for the trait, but are legal according
-    // to Rust. Any combination not in this or `allowed_combinations` is either
-    // illegal according to Rust or the behavior is unspecified. If the behavior
-    // is unspecified, it might become specified in the future, and that
-    // specification might not play nicely with our requirements. Thus, we
-    // reject combinations with unspecified behavior in addition to illegal
-    // combinations.
-    pub(crate) disallowed_but_legal_combinations: &'static [&'static [Repr]],
-}
-
-impl<R: KindRepr> Config<R> {
-    /// Validate that `input`'s representation attributes conform to the
-    /// requirements specified by this `Config`.
+/// The computed representation of a type.
+///
+/// This is the result of processing all `#[repr(...)]` attributes on a type, if
+/// any. A `Repr` is only capable of representing legal combinations of
+/// `#[repr(...)]` attributes.
+#[cfg_attr(test, derive(Copy, Clone, Debug, Eq, PartialEq))]
+pub(crate) enum Repr<Prim, Packed, S> {
```

Comment:
Why are these repr types generic? This seems like something that has a pretty concrete definition. I didn't see any code which obviously made use of the generic parameters.

Reply:
- Structs and unions cannot be `repr(Int)`, so the `Prim` type is set to `Immutable` for them (in `type StructUnionRepr`)
- Enums cannot be `repr(packed)`, so the `Packed` type is set to `Immutable` for enums (in `type EnumRepr`)
- `S` is for span information - it's `Span` normally, but stubbed as `()` in tests since `Span`s aren't easy to synthesize in testing

I could be convinced that there's a better approach for stubbing out span information - e.g. maybe I can just write the tests so that they manually ignore span information when computing equality.

Reply:
> I could be convinced that there's a better approach for stubbing out span information - e.g. maybe I can just write the tests so that they manually ignore span information when computing equality.

Did this.